### DISCUSSION TITLE
How to handle /graphql vs. REST on Github Enterprise

### DISCUSSION THREAD

**Discussion by ericscheidemantle-dreamworks** [badges/shields#6909] [Q&A] [answered]
I've found and successfully set my `GITHUB_URL` but have yet to figure out how to handle the fact that the REST API uses /api/v3 but the GRAPHQL doesn't.  If I set `GITHUB_URL` to https://myserver/api/v3, some end points like `/github/last-commit/:user/:repo` work fine, but ones using /graphql, like `/github/issues-pr/:user/:repo` fail because an invalid URL is constructed.

Is there a configuration option to solve this?

**Comment by calebcartwright** [comment]
Have you tried just setting the github url to https://my-github-enterprise-server.domain? You shouldn't have to append the api routes to the url, as that's handled by the request-making code.

Whether or not the version of Enterprise Server you are running _supports_ the v4 endpoint is a different matter altogether

**Reply by ericscheidemantle-dreamworks** [reply]
Thank-you, but it does not work w/o me including api/v3.  I get a "repo not found" badge.  It also appears, though we're on 3.0.12 of enterprise, and github documents graphql in their enterprise manual, it's not working on our end.

https://docs.github.com/en/enterprise-server@3.0/graphql/guides/forming-calls-with-graphql

**Reply by calebcartwright** [reply, marked as answer]
Ah so whereas public .com just has everything on the .api subdomain with the v4 only having an extra path part (`graphql`) server instead has an `api` path component with the rest api explicitly behind an extra `/v3`.

That's indeed the most likely source of the issue. If Enterprise Server won't accept requests on `your-github.com/api/...` for `v3` then your best bet in the short term if you need both the v3 and v4 based badges is likely to set up some redirects so that `your-github.com/api/graphql` continues on as-is, but `your-github/api/foo`... is redirected to `your-github/api/v3/foo`, and set `GITHUB_URL` in your Shields server to `your-github.com/api`

**Comment by ericscheidemantle-dreamworks** [comment]
Thank-you, I just reached a similar conclusion.  https://my-github-enterprise-server.domain/api as the GITHUB_URL works great for the graphql badges, but not the rest ones, whereas  https://my-github-enterprise-server.domain/api/v3, it's the reverse.  We'll poke around with it.
